test/test_dataloader.py

```diff
@@ -2605,6 +2605,41 @@ def test_ind_worker_queue(self):
                 self._run_ind_worker_queue_test(batch_size=batch_size, num_workers=num_workers + 1)
 
 
+class TestMultiEpochDataset(IterableDataset):
+    def __init__(self, length):
+        self.length = length
+
+    def __iter__(self):
+        worker_info = torch.utils.data.get_worker_info()
+        assert worker_info is not None
+        worker_id = worker_info.id
+        for idx in range(self.length // worker_info.num_workers):
+            if worker_id == 0:
+                time.sleep(0.001)
```

Comment:
I don't think this is needed for the test. Even without this `sleep`, the DataLoader before this PR would break this test.
The Error should be non-deterministic before your PR because it depends on which worker send `StopIteration` first.

Reply:
This line made the test fail with high probability for me that's why I added it.  Without it, you can something run multiple epochs and multiple test runs and not hit the issue.  I'm ok removing it but it will make the probability of detecting the issue lower.  On a similar note I got a timeout error in the last test run, not sure why this would cause that but I can't figure out why the test took over 5 sec to run.

Reply:
I took it out so it's fine.  One of the CI runs on an earlier commit failed with timeout.  I don't know how to get those previous logs.